_posts/2020-07-07-aws-localstack.md

```diff
@@ -0,0 +1,252 @@
+```
+
+### Execute JUnit Tests
+While running as JUnit, ***LocalStack is started as a docker container at a random port and stopped after all tests have finished execution.*** 
+
+The code snippet is a JUnit Jupiter test used to test a java class to store an object in an S3 bucket.
+
+```java
+@ExtendWith(LocalstackDockerExtension.class)
+@LocalstackDockerProperties(services = { "s3", "sqs" })
+class AwsServiceClientTest {
+
+	private static final Logger logger = Logger.getLogger(AwsServiceClient.class.getName());
+
+	private static final Region region = Region.US_EAST_1;
+	private static final String bucketName = "io.pratik.poc";
+
+	private AwsServiceClient awsServiceClient = null;
+
+	@BeforeEach
+	void setUp() throws Exception {
+		String endpoint = Localstack.INSTANCE.getEndpointS3();
+		awsServiceClient = new AwsServiceClient(endpoint);
+		createBucket();
+	}
+
+	@Test
+	void testStoreInS3() {
+		logger.info("Executing test...");
+		awsServiceClient.storeInS3("image1");
+		assertTrue(keyExistsInBucket(), "Object created");
+	}
+```
+
+
+### Spring Boot Application using S3 and DynamoDB
+
+We will create a simple customer registration application using the popular [Spring Boot](https://spring.io/projects/spring-boot) framework. Our application will have an API that will take a first name, last name, email, mobile, and a profile picture. This API will save the record in DynamoDB, and store the profile picture in an S3 bucket. 
+
+I started by creating a Spring boot rest API using https://start.spring.io with minimum dependencies for web, and Lombok. I created docker-compose.yml. The below snippet shows the environment variable SERVICES set to the name of services (s3 and dynamodb) I need to spin up for my application.
+
+```
+version: '2.1'
+
+services:
+  localstack:
+    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
+    image: localstack/localstack
+    ports:
+      - "4566-4599:4566-4599"
+      - "${PORT_WEB_UI-8080}:${PORT_WEB_UI-8080}"
+    environment:
+      - SERVICES=s3,dynamodb
+
+```
+
+I added the AWS SDK dependencies for S3 and DynamoDB to the pom.xml of the application generated in the previous step.
+
+```xml
+		<dependency>
+			<groupId>software.amazon.awssdk</groupId>
+			<artifactId>dynamodb</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>software.amazon.awssdk</groupId>
+			<artifactId>s3</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>cloud.localstack</groupId>
+			<artifactId>localstack-utils</artifactId>
+			<version>0.2.1</version>
+			<scope>test</scope>
+		</dependency>
+
+```
+I created the controller class containing the endpoint and two service classes for invoking the S3 and DynamoDB services.
+I used the default profile for real AWS services and created an additional profile named "local" for testing with LocalStack (mock AWS services). The LocalStack URL is configured in application-local.properties. Let us take a look at our CustomerProfileStore class. 
```

Comment:
Please show an `application-local.yml` and an `application.yml` file to explain the profile. Should be enough to include just the property we're using in the code.